### ISSUE TITLE
Product id sequence can overflow

### ISSUE THREAD

**Issue by whatnick** [opendatacube/datacube-explorer#144]
Cubedash gen fails in the Geoscience Australia [dev deployment](https://explorer.dev.dea.ga.gov.au/s2a_nrt_granule/2020/5/24) and presumably in production. Failure occurred about a month ago , showing the delayed NRT products in the link above however recent improvements to logging have surfaced them. The NRT streams are functional and OWS can see them :

![image](https://user-images.githubusercontent.com/491396/85236381-e0d2e880-b460-11ea-92d2-5c9fd142c8b8.png)

Failure mode is lot of messages as below per product:
```
NOTICE:  identifier "dashgen-fc-percentile-albers-seasonal agdc-1.8.1.dev22+gc2c25d3d" will be truncated to "dashgen-fc-percentile-albers-seasonal agdc-1.8.1.dev22+gc2c25d3"
2020-06-19T07:07:29.650249Z [exception] generate.product.error         product=ga_s2a_ard_nbar_granule
Traceback (most recent call last):
  File "/env/lib/python3.6/site-packages/sqlalchemy/engine/base.py", line 1284, in _execute_context
    cursor, statement, parameters, context
  File "/env/lib/python3.6/site-packages/sqlalchemy/engine/default.py", line 590, in do_execute
    cursor.execute(statement, parameters)
psycopg2.errors.SequenceGeneratorLimitExceeded: nextval: reached maximum value of sequence "product_id_seq" (32767)
The above exception was the direct cause of the following exception:
Traceback (most recent call last):
  File "/env/lib/python3.6/site-packages/cubedash/generate.py", line 46, in generate_report
    store.refresh_product(product, refresh_older_than=refresh_time)
  File "/env/lib/python3.6/site-packages/cubedash/summary/_stores.py", line 181, in refresh_product
    derived_products=derived_products,
  File "/env/lib/python3.6/site-packages/cubedash/summary/_stores.py", line 388, in _set_product_extent
    .values(name=product.name, **fields)
  File "/env/lib/python3.6/site-packages/sqlalchemy/engine/base.py", line 2244, in execute
    return connection.execute(statement, *multiparams, **params)
  File "/env/lib/python3.6/site-packages/sqlalchemy/engine/base.py", line 1020, in execute
    return meth(self, multiparams, params)
  File "/env/lib/python3.6/site-packages/sqlalchemy/sql/elements.py", line 298, in _execute_on_connection
    return connection._execute_clauseelement(self, multiparams, params)
  File "/env/lib/python3.6/site-packages/sqlalchemy/engine/base.py", line 1139, in _execute_clauseelement
    distilled_params,
  File "/env/lib/python3.6/site-packages/sqlalchemy/engine/base.py", line 1324, in _execute_context
    e, statement, parameters, cursor, context
  File "/env/lib/python3.6/site-packages/sqlalchemy/engine/base.py", line 1518, in _handle_dbapi_exception
    sqlalchemy_exception, with_traceback=exc_info[2], from_=e
  File "/env/lib/python3.6/site-packages/sqlalchemy/util/compat.py", line 178, in raise_
    raise exception
  File "/env/lib/python3.6/site-packages/sqlalchemy/engine/base.py", line 1284, in _execute_context
    cursor, statement, parameters, context
  File "/env/lib/python3.6/site-packages/sqlalchemy/engine/default.py", line 590, in do_execute
    cursor.execute(statement, parameters)
sqlalchemy.exc.DataError: (psycopg2.errors.SequenceGeneratorLimitExceeded) nextval: reached maximum value of sequence "product_id_seq" (32767)
[SQL: INSERT INTO cubedash.product (name, dataset_count, last_refresh, source_product_refs, derived_product_refs, time_earliest, time_latest) VALUES (%(name)s, %(dataset_count)s, now(), %(source_product_refs)s, %(derived_product_refs)s, %(time_earliest)s, %(time_latest)s) ON CONFLICT (name) DO UPDATE SET dataset_count = %(param_1)s, last_refresh = now(), source_product_refs = %(param_2)s, derived_product_refs = %(param_3)s, time_earliest = %(param_4)s, time_latest = %(param_5)s RETURNING cubedash.product.id]
[parameters: {'name': 'ga_s2a_ard_nbar_granule', 'dataset_count': 173779, 'source_product_refs': [], 'derived_product_refs': [], 'time_earliest': datetime.datetime(2017, 7, 2, 0, 57, 11, 26000, tzinfo=psycopg2.tz.FixedOffsetTimezone(offset=0, name=None)), 'time_latest': datetime.datetime(2019, 9, 16, 1, 17, 21, 24000, tzinfo=psycopg2.tz.FixedOffsetTimezone(offset=0, name=None)), 'param_1': 173779, 'param_2': [], 'param_3': [], 'param_4': datetime.datetime(2017, 7, 2, 0, 57, 11, 26000, tzinfo=psycopg2.tz.FixedOffsetTimezone(offset=0, name=None)), 'param_5': datetime.datetime(2019, 9, 16, 1, 17, 21, 24000, tzinfo=psycopg2.tz.FixedOffsetTimezone(offset=0, name=None))}]
(Background on this error at: http://sqlalche.me/e/9h9h)
```

This is the command the Cronjob runs :

```
- "cubedash-gen"
      - "--no-init-database"
      - "--refresh-stats"
      - "--force-refresh"
      - "--all"
```

**Comment by jeremyh** [comment]
We’re using a `SmallInt` here for the product-id, allowing a maximum of 32k products. Which matches upstream ODC. 

That’s *should* be enough, since our biggest production deployment has ~60ish products, many of which are tests or near-clones of the same data.

The underlying issue seems to be that `cubedash-gen` bumps the id counter even when no product is being added, and so running the same command via cron eventually hits the limit.

We could:
1. Bump the ID to a normal int, or
2. Fix that query so that the counter is only bumped when new products are added.

I’m leading towards the second option, in order to:
1. Not deviate from upstream ODC’s types.
2. Avoid the increase the DB row size. 
    - (The product-id is used repeatedly in statistics-arrays in individual rows such as counts-through-time, so it’s not *just* me being penny-pinching over 16-bytes. But admittedly this is not a deal-breaking concern.)

**Comment by jeremyh** [comment]
As mentioned on Slack, the short-term fix is to drop the cubedash schema/counters and recreate. Then rerun `cubedash-gen` to repopulate. If there’s still multiple instances you can do it in one instance and then switch over to avoid downtime. Otherwise we’ll need to resort to a more elaborate migration script.

**Comment by whatnick** [comment]
We have reset the sequence in dev.
```
ows=> select last_value from cubedash.product_id_seq;
 last_value 
------------
      32767
(1 row)

ows=> alter sequence cubedash.product_id_seq RESTART WITH 76;
ALTER SEQUENCE
ows=> select last_value from cubedash.product_id_seq;
 last_value 
------------
         76
(1 row)
```

**Comment by whatnick** [comment]
The max of the sequence should be greater that highest `product_id`

```
select max(id) from agdc.dataset_type;
```

**Comment by jeremyh** [comment]
It looks like @alexgleith 's fix (77a904678412b62c5dbcab3e4f775cdca193fb1d) worked, as I'm still seeing `max(id) == sequence_val` locally. Closing this unless you see other issues.
